src/backend/lib/upgrade-download.js

```diff
@@ -236,6 +237,11 @@ class Download extends EventEmitter {
           option.version,
           option.hash,
           err => {
+            // Warning: worth showing a dialog box
+            // THOUGHT: it would be nice if there was a way for this component to, like, reset itself?
+            //   some options:
+            //     1. set to state Error, like we do here, then, after a timeout, set to Idle again
+            //     2. expose a 'reset()' function that the frontend can manually call that resets the entire upgrade system
```

Comment:
I think this is an error "event" rather than an error state. As I understand it, if there is an error when downloading, then the state of the downloader returns to `IDLE`, ready to download again? e.g. and error event is what causes the state to transition from `DOWNLOADING` to `IDLE`. Or should an error event trigger a retry?

If this is implemented as an event, then the front-end can decide whether to retain a UI state for a period of time after the error occurs. I think the simplest initial implementation is to show a dialog for the error, which is displayed until the user dismisses it.

My suggestion: Do not capture this as an error state, but emit an error event when this happens.

Reply:
> As I understand it, if there is an error when downloading, then the state of the downloader returns to IDLE, ready to download again?

Currently an error in a component will put in eternally in the `"Error"` state. For the case of a failed download, we could emit an `"error"` event instead though and transition back to `"Idle"`.